### PR TITLE
feat: display node whose name length is wider than the width of nvim-tree window in floating window.

### PR DESCRIPTION
**Pull request by lu5je0**
Referencing #1297

TODO:
- ~~add an option in `require'nvim-tree'.setup`~~
- ~~show floating window when `winsaveview().leftcol` is not zero~~

### CODE REVIEW

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Doesn't look good for a right tree.

[1293-length-test.tar.gz](https://github.com/kyazdani42/nvim-tree.lua/files/8795212/1293-length-test.tar.gz)


1305-floating-node

The sign column is added to the left of the float. It should be removed, for left and right.

File: lua/nvim-tree.lua [hunk 1]
```diff
@@ -377,6 +377,7 @@ local DEFAULT_OPTS = { -- BEGIN_DEFAULT_OPTS
     add_trailing = false,
     group_empty = false,
     highlight_git = false,
+    floating_node = false,
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:380]: This option name is not intuitive; perhaps `renderer.full_name`.

File: lua/nvim-tree/renderer/components/floating-node.lua [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+local function hide(win)
+  if win then
+    if api.nvim_win_is_valid(win) then
+      api.nvim_win_close(win, true)
+    end
+  end
+end
+
+local popup_win = nil
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/components/floating-node.lua:14]: This should be a member.

File: lua/nvim-tree/renderer/components/floating-node.lua [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+local function show()
+  hide(popup_win)
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/components/floating-node.lua:16]: Is this necessary?
The first autocommand should do the job.

File: lua/nvim-tree/renderer/components/floating-node.lua [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  local line_nr = api.nvim_win_get_cursor(0)[1]
+  if line_nr == 1 then
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/components/floating-node.lua:18]: We need to call `is_root_folder_visible` to decide whether to display line 1.

**Review by alex-courtis — CHANGES_REQUESTED** [review]

**Comment by alex-courtis** [comment]
I'm not sure that this feature is necessary; #1293 achieves the goal of showing the full name and it is a lot simpler.

I do appreciate all your efforts @lu5je0

**Comment by kyazdani42** [comment]
for the people that don't want to grow / reduce the size, maybe this would be helpful anyway :)

File: lua/nvim-tree/renderer/components/floating-node.lua (hunk 4 above)
Comment (lu5je0) [reply lua/nvim-tree/renderer/components/floating-node.lua:18]: I forgeted this😄

File: lua/nvim-tree/renderer/components/floating-node.lua (hunk 3 above)
Comment (lu5je0) [reply lua/nvim-tree/renderer/components/floating-node.lua:16]: Yes, this is not necessary, I will remove it.

File: lua/nvim-tree.lua (hunk 1 above)
Comment (lu5je0) [reply lua/nvim-tree.lua:380]: looks good

**Comment by lu5je0** [comment]
> 

The popup should be hide for right tree.

**Comment by lu5je0** [comment]
> Doesn't look good for a right tree.
> 
> [1293-length-test.tar.gz](https://github.com/kyazdani42/nvim-tree.lua/files/8795212/1293-length-test.tar.gz)
> 
> 1305-floating-node [ ](https://user-images.githubusercontent.com/1505378/170912119-ed147d59-84db-41b4-8d2d-cd5bf7408931.gif) 1305-floating-node [ ](https://user-images.githubusercontent.com/1505378/170912119-ed147d59-84db-41b4-8d2d-cd5bf7408931.gif)
> 
> Edit: ~The sign column is added to the left of the float. It should be removed, for left and right.~
> 
> The popup should be left trimmed up to the icon/name.

Yes, it should be hidden for the right tree

File: lua/nvim-tree/renderer/components/floating-node.lua (hunk 2 above)
Comment (lu5je0) [reply lua/nvim-tree/renderer/components/floating-node.lua:14]: Change this to `M.popup_win`?

File: doc/nvim-tree-lua.txt [hunk 5]
```diff
@@ -475,6 +476,10 @@ UI rendering setup
     Compact folders that only contain a single folder into one node in the file tree.
       Type: `boolean`, Default: `false`
 
+    *nvim-tree.renderer.full_name*
+    Display node whose name length is wider than the width of nvim-tree window in floating window.
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:480]: Need to specify that this only applies to left trees. Maybe add:

    Only works when |nvim-tree.view.side| is `left`.

**Comment by alex-courtis** [comment]
> Yes, it should be hidden for the right tree

That works for me. Tested OK.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Looking good. Still to do:

* Solution / answer for "show floating window when winsaveview().leftcol is not zero"
* Rebase
* stylua/luacheck
*

**Comment by lu5je0** [comment]
> Solution / answer for "show floating window when winsaveview().leftcol is not zero"
There are quite complex to do this. Maybe we can just disable full name in this situation simply.

**Comment by alex-courtis** [comment]
> > Solution / answer for "show floating window when winsaveview().leftcol is not zero"
> 
> There are quite complex to do this. Maybe we can just disable full name in this situation simply.

That works for me. I'm not sure when this could happen, however it is not catastrophic if we don't show the full name.

**Review by alex-courtis — APPROVED** [review]

**Review by kyazdani42 — APPROVED** [review]

**Comment by kyazdani42** [comment]
@RaafatTurki this is unexpected behavior, although this might be caused by neovim. I've noticed some issues when playing with winbar lately. Although it works nicely most of the time.
If you can reproduce consistently, please open a new issue.

**Pull request merged**